File: agent/internal/sshutil/encoder.go
package sshutil

import (
	"encoding/binary"
	"fmt"
)

// An SSH frame is used to wrap raw binary data to and from an SSH session
// in order to allow multiplexing connections over a single file descriptor.
//
// In practice, this is how we can easily support multiple SSH connections
// through the 2 file descriptors created by the parent node process.
//
// This is the format of an SSH frame:
// - Magic: The first 4 bytes are HPLS (0x48504C53) to identify the frame.
// - Version Byte: The first byte is the version of the frame format.
// - Channel Type: The second byte indicates the type of channel.
// - Session ID: The next bytes are the length and actual session ID.
// - Payload: The remaining bytes are the payload length and actual data.
//
// +---------+----------+--------------+-------------+----------+
// | Magic   | Version  | Channel Type | SID Length  | SID      |
// | 4 bytes | 1 byte   | 1 byte       | 1 byte (S)  | S bytes  |
// +---------+----------+--------------+------------------------+
// | Payload Length     | Payload                               |
// | 4 bytes (u32, P)   | P bytes                               |
// +--------------------+---------------------------------------+

const (
	MagicString = "HPLS"
	VersionByte = 1
)

type ChannelType int

const (
	ChannelTypeStdin ChannelType = iota
	ChannelTypeStdout
	ChannelTypeStderr
)

type SSHFrame struct {
	ChannelType ChannelType
	SessionID   string
	Payload     []byte

	Length func() int
}

type HPLSFrame1 struct{}

func (t HPLSFrame1) Encode(frame SSHFrame) ([]byte, error) {
	frameChan := frame.ChannelType
	switch frameChan {
	case ChannelTypeStdin, ChannelTypeStdout, ChannelTypeStderr:
	default:
		return nil, fmt.Errorf("invalid channel type: %d", frameChan)
	}

	if len(frame.SessionID) == 0 {
		return nil, fmt.Errorf("session ID cannot be empty")
	}

	if len(frame.Payload) == 0 {
		return nil, fmt.Errorf("payload cannot be empty")
	}

	sid := []byte(frame.SessionID)
	if len(sid) > 255 {
		return nil, fmt.Errorf("session ID exceeds 255 byte limit")
	}

	if len(frame.Payload) > 0xFFFFFFFF {
		return nil, fmt.Errorf("payload exceeds 4GB limit")
	}

	frameLen := 4                      // Magic
	frameLen += 1                      // Version byte
	frameLen += 1                      // Channel type
	frameLen += 1 + len(sid)           // Session ID length + SID
	frameLen += 4 + len(frame.Payload) // Payload length + Payload

	buf := make([]byte, frameLen)
	copy(buf[0:4], []byte(MagicString))
	buf[4] = VersionByte
	buf[5] = byte(frameChan)
	buf[6] = byte(len(sid))

	offset := 7 + len(sid)
	copy(buf[7:offset], sid)

	binary.BigEndian.PutUint32(buf[offset:offset+4], uint32(len(frame.Payload)))
	copy(buf[offset+4:], frame.Payload)
	return buf, nil
}

func (t HPLSFrame1) Decode(buf []byte) (SSHFrame, error) {
	frame := SSHFrame{}
	if len(buf) < 5 || string(buf[0:4]) != MagicString || buf[4] != VersionByte {
		return frame, fmt.Errorf("illegal HPLS1 frame format")
	}

	frame.ChannelType = ChannelType(buf[5])
	if frame.ChannelType < ChannelTypeStdin || frame.ChannelType > ChannelTypeStderr {
		return frame, fmt.Errorf("invalid channel type: %d", frame.ChannelType)
	}

	sidLen := int(buf[6])
	if len(buf) < 7+sidLen+4 {
		return frame, fmt.Errorf("buffer too short for session ID and payload length")
	}

	frame.SessionID = string(buf[7 : 7+sidLen])
	payloadLen := int(binary.BigEndian.Uint32(buf[7+sidLen:]))
	if len(buf) < 7+sidLen+4+payloadLen {
		return frame, fmt.Errorf("buffer too short for payload")
	}

	frame.Payload = buf[7+sidLen+4 : 7+sidLen+4+payloadLen]
	frame.Length = func() int {
		return 7 + sidLen + 4 + payloadLen
	}

	return frame, nil
}

File: app/server/agent/encoder.ts
// Refer to agent/internal/sshutil/encoder.go for more details
// This is the Node.js implementation of the SSH encoder
import log from '~/utils/log';

const MAGIC = 'HPLS';
const VERSION = 1;

// 0 -> Stdin
// 1 -> Stdout
// 2 -> Stderr
export type ChannelType = 0 | 1 | 2;

interface SSHFrame {
	sessionId: string;
	channel: ChannelType;
	payload: Blob | ArrayBufferLike | string;
}

export async function encodeSSHFrame(frame: SSHFrame) {
	const sid = Buffer.from(frame.sessionId, 'utf8');
	if (sid.length > 255) {
		log.error('agent', 'SSH session ID too long: %s', frame.sessionId);
		return;
	}

	const payload = Buffer.isBuffer(frame.payload)
		? frame.payload
		: typeof frame.payload === 'string'
			? Buffer.from(frame.payload, 'utf8')
			: frame.payload instanceof Blob
				? Buffer.from(await frame.payload.arrayBuffer())
				: Buffer.from(frame.payload);

	// Size can only hold 4 bytes
	if (payload.length > 0xffffffff) {
		log.error('agent', 'SSH payload too large: %d bytes', payload.length);
		return;
	}

	const frameSize =
		4 + // Magic
		1 + // Version
		1 + // Channel Type
		(1 + sid.length) + // Session ID length + SID
		(4 + payload.length); // Payload length + Payload

	const buf = Buffer.alloc(frameSize);
	buf.write(MAGIC, 0, 'utf8');
	buf.writeUInt8(VERSION, 4);
	buf.writeUInt8(frame.channel, 5);
	buf.writeUInt8(sid.length, 6);

	const offset = 7 + sid.length;
	sid.copy(buf, 7);

	buf.writeUInt32BE(payload.length, offset);
	payload.copy(buf, offset + 4);
	return buf;
}

export function decodeSSHFrame(data: Buffer) {
	if (data.length < 5) {
		log.error('agent', 'SSH frame too short: %d bytes', data.length);
		return;
	}

	const magic = data.toString('utf8', 0, 4);
	const version = data.readUInt8(4);

	if (magic !== MAGIC || version !== VERSION) {
		log.error('agent', 'Invalid SSH frame magic or version');
		return;
	}

	const channel = data.readUInt8(5) as ChannelType;
	const sidLength = data.readUInt8(6);
	if (data.length < 7 + sidLength + 4) {
		log.error('agent', 'SSH frame too short for session ID and payload');
		return;
	}

	const sessionId = data.toString('utf8', 7, 7 + sidLength);
	const payloadLength = data.readUInt32BE(7 + sidLength);
	if (data.length < 7 + sidLength + 4 + payloadLength) {
		log.error('agent', 'SSH frame too short for payload');
		return;
	}

	const payload = data.subarray(
		7 + sidLength + 4,
		7 + sidLength + 4 + payloadLength,
	);

	return {
		sessionId,
		channel,
		payload,
	};
}
